blog/git/1 git blame.md

```diff
@@ -0,0 +1,141 @@
+commit 2ab752fcb792a742ac0fda9ddb4d65a97045090d
+Add line 7                             
+...            
+```
+We're in luck, ```git bisect``` is interactive, so let's let it guide us through this rabbit hole. Once you start you will be in bisect mode untill you type ```git bisect reset```.
+```shell 
+git bisect
+```
+This command has no output. Let's give it some input.
+> You can use short commit IDs but we already have the long ones.
+```shell
+git bisect bad 778d62a21ae08ff38db34bf9465db423f7e990fb
+git bisect good 6e83eec542f915bdaeb9565fe5dbd1c6cfaec4d6
+```
+And it will output a useful number of revisions and a calculated distance. With longer histories you'll see the power of binary search.
+
+```shell 
+Bisecting: 4 revisions left to test after this (roughly 2 steps)
+[219e531c50779854789dbae70a911dbfb2c652e8] Add line 9
+```
+Now, the second part is the interesting one. Notice how a commit that we didn't input was printed? Well, now it is checked out and we can test it to know if it's a *good* or *bad* one.
+```shell 
+cat wm-git-blame.txt
+```
+We can see that this specific commit, even if incomplete, has the correct lyrics, so we'll tell git this one is good.
+```shell 
+git bisect good
+```
+And now the output will be similar, also checking out another commit. 
+```shell 
+Bisecting: 2 revisions left to test after this (roughly 1 step)
+[e1249a7f0d86a8dca982289aa35e3892fe189ae2] change something for nothing
+```
+We check again.
+```shell 
+cat wm-git-blame.txt
+```
+*Aha!* We have found the error, or bug in a real development environment.
+```shell 
+git bisect bad
+```
+Outpust tells us that this is the last step.
+```shell 
+Bisecting: 0 revisions left to test after this (roughly 0 steps)
+[c93b967c5d5e7a601e83c3adda2b3ab814ad10f1] Add line 10
+```
+We check for the last time.
+```shell 
+cat wm-git-blame.txt
+```
+No error so...
+```shell 
+git bisect good
+```
+And now git tells us the offending commit has been found, the commit author, date, commit message and lists all changed files.
+```shell 
+e1249a7f0d86a8dca982289aa35e3892fe189ae2 is the first bad commit
+commit e1249a7f0d86a8dca982289aa35e3892fe189ae2
+Author: estebandalelr <estebandalelr@gmail.com>
+Date:   Mon Mar 7 11:07:06 2022 -0600
+
+    change something for nothing
+
+:100644 100644 ebecac7931b48bb4aaa63e1e2a8624712d959eb2 ffc445f53b954c6ad8c05a8a860920df064388d1 M      wm-git-blame.txt
+```
+So we close the case, it was me (as it's common), we can fix it commit and close the Jira ticket.
+```shell 
+git bisect reset
+```
+Now we are in the commit we were in (usually HEAD) when we started, and know what broke our app.
+
+Phew, that was a lot. Coding as a team is hard, but if you do it right it doesn't have to be crazy. The right tools help, and that's why we built [Watermelon](https://watermelon.tools), a VSCode extension that gets the historical context of your code to make it easier. For now we have github, with Jira, Zoom and Slack integrations coming soon.
```

Comment:
Too salesy. Please study how o[ther young open-source companies](https://posthog.com/blog/hightouch-posthog-reverse-etl-integration) do it.